streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java

```diff
@@ -447,11 +453,106 @@ public String getStoreForChangelogTopic(final String topicName) {
     }
 
     void addSubscribedTopicsFromMetadata(final Set<String> topics, final String logPrefix) {
-        applyToEachBuilder(b -> b.addSubscribedTopicsFromMetadata(topics, logPrefix));
+        if (usesPatternSubscription()) {
+            final Map<String, Set<String>> newTopicsByTopology = new HashMap<>();
+            final Set<String> duplicateInputTopics = new HashSet<>();
+            final Set<String> subscriptionTopicsWithUnknownSource = new HashSet<>();
+            for (final String topic : topics) {
+                final Set<String> subscribingTopologies = new HashSet<>();
+                applyToEachBuilder(b -> {
+                    if (b.sourceTopicCollection().contains(topic) || b.matchesSubscribedPattern(topic)) {
+                        subscribingTopologies.add(getTopologyNameOrElseUnnamed(b.topologyName()));
+                    }
+                });
+                if (subscribingTopologies.size() > 1) {
+                    log.error("{}Subscribed topic {} matches more than one topology: {}",
+                        logPrefix, topic, subscribingTopologies);
+                    duplicateInputTopics.add(topic);
+                } else if (subscribingTopologies.isEmpty()) {
+                    log.error("{}Topic {} is subscribed to by the consumer, but no topology can be identified " +
+                        "that is subscribing to that topic or to a matching Pattern", logPrefix, topic);
+                    subscriptionTopicsWithUnknownSource.add(topic);
+                }
+                for (final String topology : subscribingTopologies) {
+                    if (!allInputTopics.contains(topic)) {
+                        newTopicsByTopology
+                            .computeIfAbsent(topology, t -> new HashSet<>())
+                            .add(topic);
+                    }
+                }
+            }
+
+            if (!duplicateInputTopics.isEmpty()) {
+                throw new IllegalStateException("The following topics are subscribed to " +
+                    "by multiple topologies: " + duplicateInputTopics);
+            } else if (subscriptionTopicsWithUnknownSource.isEmpty()) {
+                throw new IllegalStateException("The following topics appear in the subscription " +
```

Comment:
You might already be on it, but can we add tests for both these cases?

Reply:
Yep, just haven't had time to do tests for this yet -- clearly they're needed 🙂

Reply:
Should we also check that the `newTopicsByTopology.keySet` are all new topologies, i.e. they do not have existing topics?